File: android/app/src/main/java/id/quezacolt/veilkeeper/crypto/Hkdf.kt
package id.quezacolt.veilkeeper.crypto

import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

/**
 * HKDF-SHA256 (RFC 5869), used to derive the domain-separated AuthKey and
 * WrapKey subkeys from MasterKey (CLAUDE.md Resolved Design Decision #1).
 *
 * Not available as a ready-made primitive in the Android/JDK standard
 * library (unlike AES-GCM via [javax.crypto.Cipher] or HMAC via
 * [javax.crypto.Mac], which this *is* built on), but HMAC-SHA256 is --
 * per SPEC-BASE.md Section 56 Rule 7 ("verify whether the ... standard
 * library ... can solve the problem" before adding a dependency), a ~30
 * line RFC-faithful implementation over stdlib Mac is preferable to pulling
 * in a general-purpose crypto library (e.g. Tink) for one function.
 */
object Hkdf {
    private const val HMAC_ALGORITHM = "HmacSHA256"
    private const val HASH_LENGTH = 32

    /**
     * HKDF-Extract followed by HKDF-Expand, producing [outputLength] bytes.
     *
     * @param ikm Input keying material (here: MasterKey, already
     *            high-entropy from Argon2id, so an all-zero salt is used
     *            for Extract -- standard practice per RFC 5869 Section 3.1
     *            when a salt value isn't otherwise available/needed for
     *            additional entropy concentration).
     * @param info Domain-separation context (e.g. "veilkeeper:auth:v1").
     */
    fun deriveKey(ikm: ByteArray, info: ByteArray, outputLength: Int = HASH_LENGTH): ByteArray {
        val salt = ByteArray(HASH_LENGTH) // zero-filled, see doc comment above
        val prk = extract(salt, ikm)
        return expand(prk, info, outputLength)
    }

    private fun extract(salt: ByteArray, ikm: ByteArray): ByteArray {
        val mac = Mac.getInstance(HMAC_ALGORITHM)
        mac.init(SecretKeySpec(salt, HMAC_ALGORITHM))
        return mac.doFinal(ikm)
    }

    private fun expand(prk: ByteArray, info: ByteArray, outputLength: Int): ByteArray {
        require(outputLength in 1..(255 * HASH_LENGTH)) { "HKDF: invalid output length $outputLength" }

        val mac = Mac.getInstance(HMAC_ALGORITHM)
        mac.init(SecretKeySpec(prk, HMAC_ALGORITHM))

        val output = ByteArray(outputLength)
        var previousBlock = ByteArray(0)
        var generated = 0
        var counter = 1

        while (generated < outputLength) {
            mac.reset()
            mac.update(previousBlock)
            mac.update(info)
            mac.update(counter.toByte())
            val block = mac.doFinal()

            val toCopy = minOf(block.size, outputLength - generated)
            System.arraycopy(block, 0, output, generated, toCopy)

            generated += toCopy
            previousBlock = block
            counter++
        }

        return output
    }
}

File: web/src/crypto/hkdf.ts
/**
 * HKDF-SHA256 (RFC 5869), used to derive the domain-separated AuthKey and
 * WrapKey subkeys from MasterKey (CLAUDE.md Resolved Design Decision #1).
 *
 * Implemented via the native Web Crypto API (`crypto.subtle`, algorithm
 * `"HKDF"`) -- browsers support HKDF as a first-class `deriveBits`
 * algorithm, so unlike Android (which hand-rolls it over `javax.crypto.Mac`
 * because the JDK/Android stdlib has no ready-made HKDF), Web needs no
 * hand-rolled implementation here.
 *
 * IMPORTANT for cross-platform parity: RFC 5869 Section 2.2 says that when
 * no salt is provided, HKDF-Extract uses a salt of `HashLen` (32 for
 * SHA-256) zero bytes. The Web Crypto HKDF algorithm does NOT apply that
 * default itself -- an omitted/empty `salt` there is used as a genuine
 * zero-length HMAC key, which is *not* the same as a 32-byte zero-filled
 * key. Android's `Hkdf.kt` always passes an explicit 32-byte zero salt, so
 * this implementation does too, to guarantee byte-identical output for the
 * same (ikm, info, outputLength).
 */
const HASH_LENGTH_BYTES = 32

export async function deriveKey(
  ikm: Uint8Array,
  info: Uint8Array,
  outputLength: number = HASH_LENGTH_BYTES,
): Promise<Uint8Array> {
  const salt = new Uint8Array(HASH_LENGTH_BYTES) // zero-filled, see doc comment above

  const key = await crypto.subtle.importKey('raw', toArrayBuffer(ikm), 'HKDF', false, ['deriveBits'])

  const bits = await crypto.subtle.deriveBits(
    {
      name: 'HKDF',
      hash: 'SHA-256',
      salt: toArrayBuffer(salt),
      info: toArrayBuffer(info),
    },
    key,
    outputLength * 8,
  )

  return new Uint8Array(bits)
}

function toArrayBuffer(bytes: Uint8Array): ArrayBuffer {
  // Slice to guarantee a plain ArrayBuffer backing (not SharedArrayBuffer,
  // and not a view into a larger buffer), which is what crypto.subtle
  // expects/accepts across all supported browsers.
  return bytes.slice().buffer
}
